src/meta/src/stream/stream_manager.rs

```diff
@@ -166,39 +168,36 @@ where
                 &mut self,
                 stream_node: &mut StreamNode,
                 actor_id: ActorId,
+                vnode_mapping: &Option<Buffer>,
                 same_worker_node_as_upstream: bool,
                 is_singleton: bool,
             ) -> MetaResult<()> {
                 let Some(NodeBody::Chain(ref mut chain)) = stream_node.node_body else {
                     // If node is not chain node, recursively deal with input nodes
                     for input in &mut stream_node.input {
-                        self.resolve_chain_node_inner(input, actor_id, same_worker_node_as_upstream, is_singleton)?;
+                        self.resolve_chain_node_inner(input, actor_id, vnode_mapping, same_worker_node_as_upstream, is_singleton)?;
                     }
                     return Ok(());
                 };
 
                 // get upstream table id
                 let table_id = TableId::new(chain.table_id);
-
-                // FIXME: We assume that the chain node is always on the same parallel unit as its
-                // upstream materialize node here to find the upstream actor.
                 let upstream_actor_id = {
-                    // 1. use table id to get upstream parallel_unit -> actor_id mapping
-                    let upstream_parallel_actor_mapping =
-                        &self.upstream_parallel_unit_info[&table_id];
+                    // 1. use table id to get upstream vnode mapping info: [(actor_id,
+                    // option(vnode_mapping))]
+                    let upstream_vnode_mapping_info = &self.upstream_vnode_mapping_info[&table_id];
 
                     if is_singleton {
                         // Directly find the singleton actor id.
-                        *upstream_parallel_actor_mapping
-                            .values()
-                            .exactly_one()
-                            .unwrap()
+                        upstream_vnode_mapping_info.iter().exactly_one().unwrap().0
                     } else {
-                        // 2. use our actor id to get parallel unit id of the chain actor
-                        let parallel_unit_id = self.locations.actor_locations[&actor_id].id;
-                        // 3. and use chain actor's parallel unit id to get the corresponding
-                        // upstream actor id
-                        upstream_parallel_actor_mapping[&parallel_unit_id]
+                        // 2. find the upstream actor id by vnode mapping.
+                        assert!(vnode_mapping.is_some());
+                        upstream_vnode_mapping_info
+                            .iter()
+                            .find(|(_, bitmap)| bitmap == vnode_mapping)
```

Comment:
I think we should still use the parallel unit id to find the upstream actor, and follow the upstream bitmap. Note that the vnode bitmap of this chain fragment is generated by us, while the upstream materialize may have been scaled. Even if they have the same parallelism, they'll tend not to be the same so we'll fail to find it here.

For example,
1. The upstream initially has 3 parallelisms, the mapping is `[1,1,1,1,2,2,2,2,3,3,3,3]`.
2. The upstream scales. To achieve better locality, the mapping becomes `[1,1,1,4,2,2,2,4,3,3,3,4]`.
3. We schedule a new MV-on-MV with 4 parallelisms initially, so the mapping of the chain is `[1,1,1,2,2,2,3,3,3,4,4,4]` by default.

Reply:
Well, current schedule strategy can't cover this case. I decided to ignore chain actor when scheduling and setup its upstream and vnode mapping when resolving chain node while creating. This will cover the mentioned case and other scaling scenario.